### PR TITLE
Python 3 support and CI

### PR DESCRIPTION
**Pull request by lasote**
Ready, tests OK in Python 2.7, 3.3, 3.4 and 3.5.
I suggest a review and merge to try it hard for the next release.

### CODE REVIEW

File: conans/client/rest/rest_client.py [hunk 1]
```diff
@@ -109,7 +110,7 @@ def get_conanfile(self, conan_reference):
         url = "%s/conans/%s/download_urls" % (self._remote_api_url, "/".join(conan_reference))
         urls = self._get_json(url)
 
-        if CONANFILE not in urls.keys():
+        if CONANFILE not in list(urls.keys()):
```
Comment (memsharded) [inline conans/client/rest/rest_client.py]: It seems that this could be ```list(urls)```

File: conans/client/loader.py [hunk 2]
```diff
@@ -68,7 +74,7 @@ def load_conan(self, conan_file_path, output, consumer=False):
         try:
             current_dir = os.path.dirname(conan_file_path)
             sys.path.append(current_dir)
-            old_modules = sys.modules.keys()
+            old_modules = list(sys.modules.keys())
```
Comment (memsharded) [inline conans/client/loader.py]: Probably this could be ```list(sys.modules)```

File: conans/server/conf/__init__.py [hunk 3]
```diff
@@ -139,8 +139,8 @@ def write_permissions(self):
     def users(self):
         def validate_pass_encoding(password):
             try:
-                password.encode('latin1')
-            except UnicodeDecodeError:
+                password.encode('ascii')
```
Comment (memsharded) [inline conans/server/conf/__init__.py]: Sure that this change from latin1 to ascii will not have problems for users with passwords with latin1 only chars?

File: conans/client/rest/uploader_downloader.py [hunk 4]
```diff
@@ -11,8 +11,7 @@ def __init__(self, requester, output, verify, chunk_size=1000):
 
     def post(self, url, content):
         self.output.info("")
-        it = upload_in_chunks(content, self.chunk_size, self.output)
-        return self.requester.put(url, data=IterableToFileAdapter(it), verify=self.verify)
+        return self.requester.put(url, data=content, verify=self.verify)
```
Comment (memsharded) [inline conans/client/rest/uploader_downloader.py]: TODO: check if not having upload progress bar is an issue

File: conans/tools.py [hunk 5]
```diff
@@ -42,17 +42,17 @@ def unzip(filename, destination="."):
     full_path = os.path.normpath(os.path.join(os.getcwd(), destination))
     with zipfile.ZipFile(filename, "r") as z:
         uncompress_size = sum((file_.file_size for file_ in z.infolist()))
-        print "Unzipping %s, this can take a while" % (human_size(uncompress_size))
+        print("Unzipping %s, this can take a while" % human_size(uncompress_size))
         extracted_size = 0
         for file_ in z.infolist():
             extracted_size += file_.file_size
-            print "Unzipping %.0f %%\r" % (extracted_size * 100.0 / uncompress_size),
+            print("Unzipping %.0f %%\r" % (extracted_size * 100.0 / uncompress_size),)
```
Comment (memsharded) [inline conans/tools.py]: This trailing comma doesn't have the same meaning that in Py2 without parenthesis

File: conans/util/env_reader.py [hunk 6]
```diff
@@ -6,21 +6,28 @@
...
+                     int: lambda x: int(x),
+                     float: lambda x: float(x),
+                     list: lambda x: x.split(","),
+                     bool: lambda x: x == '1'}
 
     env_var = environment.get(env_key, default)
     if env_var != default:
-        func = _default_type[type(default)]
-        env_var = func(env_var)
+        if isinstance(default, str):
```
Comment (memsharded) [inline conans/util/env_reader.py]: what fails to not use the already defined lambda?

**Comment by memsharded** [comment]
I'd say that overall it is a great job. A few concerns that maybe should be considered:

- Latin1 characters in already existing passwords
- Lack of upload progress bar (uploads from ADSL are typically slower)
- Trailing comma in print() in unzip (will produce lots of output)

File: conans/server/conf/__init__.py (hunk 3 above)
Comment (lasote) [inline conans/server/conf/__init__.py]: In the website we limit to ascii

File: conans/tools.py (hunk 5 above)
Comment (lasote) [inline conans/tools.py]: Oh, please check it

File: conans/util/env_reader.py (hunk 6 above)
Comment (lasote) [inline conans/util/env_reader.py]: The question is what is not failing ;) Nonetype doesn't exist anymore and can't remember well, but there was some other thing failing, maybe the types indexing....

**Pull request merged**